README.md
# Selenium::Remote::Driver [![Build Status](https://travis-ci.org/teodesian/Selenium-Remote-Driver.svg?branch=master)](https://travis-ci.org/teodesian/Selenium-Remote-Driver)
[![Coverage Status](https://coveralls.io/repos/github/teodesian/Selenium-Remote-Driver/badge.svg?branch=master)](https://coveralls.io/github/teodesian/Selenium-Remote-Driver?branch=master)

[Selenium WebDriver][wd] is a test tool that allows you to write
automated web application UI tests in any programming language against
any HTTP website using any mainstream JavaScript-enabled browser. This
module is a Perl implementation of the client for the Webdriver
[JSONWireProtocol that Selenium provides.][jsonwire]

[wd]: http://www.seleniumhq.org/
[jsonwire]: https://code.google.com/p/selenium/wiki/JsonWireProtocol
[standalone]: http://selenium-release.storage.googleapis.com/index.html

## Installation

It's probably easiest to use the `cpanm` or `CPAN` commands:

```bash
$ cpanm Selenium::Remote::Driver
```

If you want to install from this repository, see the
[installation docs][] for more details.

[installation docs]: /INSTALL.md

## Usage

You can use this module to directly start the webdriver servers, after
[downloading the appropriate ones][dl] and putting the servers in your
`$PATH`. This method does not require the JRE/JDK to be installed, nor
does it require the standalone server jar, despite the name of the
module. In this case, you'll want to use the appropriate class for
driver construction: either [Selenium::Chrome][],
[Selenium::Firefox][], [Selenium::PhantomJS][], or
[Selenium::InternetExplorer][].

You can also use this module with the `selenium-standalone-server.jar`
to let it handle browser start up for you, and also manage Remote
connections where the server jar is not running on the same machine as
your test script is executing. The main class for this method is
[Selenium::Remote::Driver][].

Regardless of which method you use to construct your browser object,
all of the classes use the functions listed in the S::R::Driver POD
documentation, so interacting with the browser, the page, and its
elements would be the same.

[Selenium::Firefox]: https://metacpan.org/pod/Selenium::Firefox
[Selenium::Chrome]: https://metacpan.org/pod/Selenium::Chrome
[Selenium::PhantomJS]: https://metacpan.org/pod/Selenium::PhantomJS
[Selenium::InternetExplorer]: https://metacpan.org/pod/Selenium::InternetExplorer
[Selenium::Remote::Driver]: https://metacpan.org/pod/Selenium::Remote::Driver
[dl]: #no-standalone-server

### no standalone server

- _Firefox 48 & newer_: install the Firefox browser, download
  [geckodriver][gd] and [put it in your `$PATH`][fxpath]. If the
  Firefox browser binary is not in the default place for your OS and
  we cannot locate it via `which`, you may have to specify the binary
  location during startup. We also will need to locate the Firefox
  browser; if the Firefox browser isn't in the default location, you
  must provide it during startup in the `firefox_binary` attr.

- _Firefox 47 & older_: install the Firefox browser in the default
  place for your OS. If the Firefox browser binary is not in the
  default place for your OS, you may have to specify the
  `firefox_binary` constructor option during startup.

- _Chrome_: install the Chrome browser, [download Chromedriver][dcd]
  and get `chromedriver` in your `$PATH`.

- _PhantomJS_: install the PhantomJS binary and get `phantomjs` in
  your `$PATH`. The driver for PhantomJS, Ghostdriver, is bundled with
  PhantomJS.

When the browser(s) are installed and you have the appropriate binary
in your path, you should be able to do the following:

```perl
my $firefox = Selenium::Firefox->new;
$firefox->get('http://www.google.com');

my $chrome = Selenium::Chrome->new;
$chrome->get('http://www.google.com');

my $ghost = Selenium::PhantomJS->new;
$ghost->get('http://www.google.com');
```

Note that you can also pass a `binary` argument to any of the above
classes to manually specify what binary to start. Note that this
`binary` refers to the driver server, _not_ the browser executable.

```perl
my $chrome = Selenium::Chrome->new(binary => '~/Downloads/chromedriver');
```

See the pod for the different modules for more details.

[dcd]: https://sites.google.com/a/chromium.org/chromedriver/downloads
[fxpath]: https://developer.mozilla.org/en-US/docs/Mozilla/QA/Marionette/WebDriver#Add_executable_to_system_path
[gd]: https://github.com/mozilla/geckodriver/releases

#### Breaking Changes for Selenium::Firefox in v1.0+

There are breaking changes for Selenium::Firefox from v0.2701 of S:F
to v1.0+. This is because in FF47 and older, Firefox didn't have a
separate webdriver server executable - local startup was accomplished
by starting your actual Firefox browser with a webdriver
extension. However, in FF48 and newer, Mozilla have switched to using
`geckodriver` to handle the Webdriver communication. Accordingly,
v1.0+ of Selenium::Firefox assumes the geckodriver setup which only
works for FF48 and higher:

```perl
# marionette_enabled defaults to 1 === assumes you're running FF48
my $fx48 = Selenium::Firefox->new;
my $fx48 = Selenium::Firefox->new( marionette_enabled => 1 );
```

To drive FF47 with v1.0+ of Selenium::Firefox, you must manually
disable marionette:

```perl
my $fx47 = Selenium::Firefox->new( marionette_enabled => 0 );
```

Doing so will start up your Firefox browser with the webdriver
extension. Note that in our tests, doing the old
"webdriver-extension-startup" for Firefox 48 does not work. Likewise,
`geckodriver` does not work with FF47.

### with a standalone server

Download the [standalone server][] and have it running on port 4444:

    $ java -jar selenium-server-standalone-X.XX.X.jar

As before, have the browsers themselves installed on your machine, and
download the appropriate binary server, passing its location to the
server jar during startup.

[standalone server]: http://selenium-release.storage.googleapis.com/index.html

#### Locally

```perl
use strict;
use warnings;
use Selenium::Remote::Driver;

my $driver = Selenium::Remote::Driver->new;
$driver->get('http://www.google.com');
print $driver->get_title . "\n"; # "Google"

my $query = $driver->find_element('q', 'name');
$query->send_keys('CPAN Selenium Remote Driver');

my $send_search = $driver->find_element('btnG', 'name');
$send_search->click;

# make the find_element blocking for a second to allow the title to change
$driver->set_implicit_wait_timeout(2000);
my $results = $driver->find_element('search', 'id');

print $driver->get_title . "\n"; # CPAN Selenium Remote Driver - Google Search
$driver->quit;
```

#### Saucelabs

If using Saucelabs, there's no need to have the standalone server
running on a local port, since Saucelabs provides it.

```perl
use Selenium::Remote::Driver;

my $user = $ENV{SAUCE_USERNAME};
my $key = $ENV{SAUCE_ACCESS_KEY};

my $driver = Selenium::Remote::Driver->new(
    remote_server_addr => $user . ':' . $key . '@ondemand.saucelabs.com',
    port => 80
);

$driver->get('http://www.google.com');
print $driver->get_title();
$driver->quit();
```

There are additional usage examples on [metacpan][meta], and also
[in this project's wiki][wiki], including
[setting up the standalone server][setup], running tests on
[Internet Explorer][ie], [Chrome][chrome], [PhantomJS][pjs], and other
useful [example snippets][ex].

[wiki]: https://github.com/teodesian/Selenium-Remote-Driver/wiki
[setup]: https://github.com/teodesian/Selenium-Remote-Driver/wiki/Getting-Started-with-Selenium%3A%3ARemote%3A%3ADriver
[ie]: https://github.com/teodesian/Selenium-Remote-Driver/wiki/IE-browser-automation
[chrome]: https://github.com/teodesian/Selenium-Remote-Driver/wiki/Chrome-browser-automation
[pjs]: https://github.com/teodesian/Selenium-Remote-Driver/wiki/PhantomJS-Headless-Browser-Automation
[ex]: https://github.com/teodesian/Selenium-Remote-Driver/wiki/Example-Snippets

## Selenium IDE Plugin

[ide-plugin.js](./ide-plugin.js) is a Selenium IDE Plugin which allows
you to export tests recorded in Selenium IDE to a perl script.

### Installation in Selenium IDE

  1. Open Selenium IDE
  2. Options >> Options
  3. Formats Tab
  4. Click Add at the bottom
  5. In the name field call it 'Perl-Webdriver'
  6. Paste this entire source in the main textbox
  7. Click 'Save'
  8. Click 'Ok'
  9. Close Selenium IDE and open it again.

**Note:** The Code generated by this plugin currently assumes that you have firefox 48 or newer + the geckodriver plugin installed.

If you have an older version of firefox you can adjust the plugin the following way:
In the options window of selenium IDE (Formats tab) there is a input box called "Header".
Here you can modify the fixed header code beeing generatd. Change Selenium::Firefox->new();
according to the firefox init examples shown above in this README.

## Support and Documentation

There is a mailing list available at

https://groups.google.com/forum/#!forum/selenium-remote-driver

for usage questions and ensuing discussions. If you've come across a
bug, please open an issue in the [Github issue tracker][issue]. The
POD is available in the usual places, including [metacpan][meta], and
in your shell via `perldoc`.

```bash
$ perldoc Selenium::Remote::Driver
$ perldoc Selenium::Remote::WebElement
```

[issue]: https://github.com/teodesian/Selenium-Remote-Driver/issues
[meta]: https://metacpan.org/pod/Selenium::Remote::Driver

## Contributing

Thanks for considering contributing! The contributing guidelines are
[in the wiki][contrib]. The documentation there also includes
information on generating new Linux recordings for Travis.

[contrib]: https://github.com/teodesian/Selenium-Remote-Driver/wiki/Contribution-Guide

## Copyright and License

Copyright (c) 2010-2011 Aditya Ivaturi, Gordon Child

Copyright (c) 2014-2017 Daniel Gempesaw

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


Changes
Revision history for Selenium-Remote-Driver

1.00   04-08-2016
        (bundled firefox driver version: 3.0.0b2)

        [BREAKING]
        - For users of Selenium::Firefox, FF support defaults to v48;
          if you have 47 installed, your existing scripts will require
          modification. See docs in Selenium::Firefox for more info.

        [NEW FEATURES]
        - #271: Support geckodriver & FF48
        - #260: Update firefox artifacts to 3.0.0b2 to support FF47
        - #259: Allow users to pass in existing FF profile dir
        - #276: Simplify recordings to enable easier releases

        [BUG FIXES]
        - Encode Firefox profiles without newlines for Sel 3.0 compatibility
        - #272: Remove unnecessary sleeps for successful waits
        - #270: Fix bug in find_child_element default finder
        - #263, #275: Address various documentation issues
        - #253: Fix error_handler being called incorrectly


0.2701 01-11-2016
        (firefox driver version: 2.48.2)

        [BUG FIXES]
        - #239: Respect existing prefs when using custom profile with Selenium::Firefox
        - #238: Update firefox driver extension to latest 2.48.2
        - #21: Update docs about using child finders with xpath

0.27   08-30-2015
        [NEW FEATURES]
        - #219: Add `startup_timeout` to CanStartBinary role
        - #211: @peroumal1: Document locator usage for Test::SRD functions

        [BUG FIXES]
        - #215: Replace Time::Mock with the more reliable & tested Test::Time
        - #212: @peroumal1: Fix bug in find_no_element_ok
        - #209: @peroumal1: Make error_handler catch non-webdriver exceptions in Test::SRD
        - #222: Describe demolish/destroy behavior.
        - #226: Allow local file URLs for PAC url
        - Stop trying to do things during global destruction

0.26   05-16-2015
        (selenium version: 2.45.0)

        [NEW FEATURES]
        - #203: Allow users to install a custom error handler
        - #204: Implement action chains

        [BUG FIXES]
        - #147: Allow for inflation to a full Moose class
        - #192: Stop assuming objects are around during global destruction
        - #200: Fix undocumented upload_file endpoint bug for paths with ".."
        - Refactor to remove some unnecessary dependencies

0.25   03-24-2015
        (selenium version: 2.45.0)

        [NEW FEATURES]
        - #189 Start webdrivers directly, removing the need for the JRE. See
           Selenium::Chrome, Selenium::Firefox, Selenium::PhantomJS.

        [BUG FIXES]
        - #194, #195: @teodesian, @jamadam: fix undocumented upload file behavior

0.24    02-18-2015
        [NEW FEATURES]
        - #182 Non-crashing warn-instead-of-die parameter finders: find_element_by_class, etc
        - #183 Convenience packages for starting up browsers: Selenium::Chrome, Selenium::Firefox, etc
        - Provide wait_until utility functionality in Selenium::Waiter

        [BUG FIXES]
        - #184 Fix bug about reverting default finder in T:S:R:D

0.23   01-25-2014
        [NEW FEATURES]
        - #178 Fix upload_file's return value & add test suite
        - #179 Add optional middle finder-strategy argument to T:S:R:D's find_element-like functions
        - #174 Add new endpoints: cache status, geolocation, log types, orientation, etc

        [BUG FIXES]
        - #175 Fix find_elements
        - #180 Stop overwriting body_text_unlike test descriptions
        - #141 Explicitly cast height and weight to integers for set_window_size

0.2203 01-18-2014
        [BUG FIXES]
        - #175 Fix find_elements_ok

0.2202 11-26-2014
        [BUG FIXES]
        - Generalize error message for invalid finder strategies

0.2201 11-13-2014
        [BUG FIXES]
        - Fix redirect handling mishap in RemoteConnection

0.22   11-10-2014
        [NEW FEATURES]
        - Improved error messaging for Saucelabs non-json cases
        - Improved error messaging for malconfigured chromedrivers
        - #169 Add maximize_window function
        - #58 Add get_user_agent function

        [MAINTENAINCE]
        - Specify minimum compatible version of Moo dep
        - Make Driver.pm more easily subclassed ( ea68e06, 63e78e3 )
        - Get TravisCI passing again
        - #164 Roll & use our own mocking classes
        - #152 Gracefully catch more error cases

        [BREAKING]
        - #161 get|set_speed subs are no-ops and will be removed in
          the upcoming release
        - #158 WebElement's drag function is deprecated and will be
          removed in the upcoming release. Use mouse_move_to_location
          in the interim

0.2102 7-16-2014
        [NEW FEATURES]
        - Add inner_window_size sub and hash key option to accomodate S-R-D-UserAgent use

0.2101 7-16-2014
        [BUG FIXES]
        - Update list of no_content_success methods that interpret an undef value as success

0.21   7-16-2014
        [NEW FEATURES]
        - #87: Implement is_hidden for WebElements

        [BUG FIXES]
        - #145: Revert previous 2.42.2 fix to allow for endpoints properly returning ''
        - #146: Keep test description when using content_like

0.2002 7-2-2014
        [BUG FIXES]
        - #135, #136, #142: Fix compatability issues with JSONWireProtocol changes in 2.42.2

        [NEW FEATURES]
        - Add an optional base_url parameter for automatic concatenation in get/get_ok

0.2001 5-23-2014
        [BUG FIXES]
        - #128: Fix upload_file where encode_base64 was not exported properly
        - #54 : Update the status endpoint when checking for Grid
        - #124: Explicitly specify finder type in get_body

0.20   4-25-2014
        [DESIRED CAPABILITIES]
        - Add a desired_capabilities option that provides full control over the desiredCapability object
        - Add a new_from_caps constructor to distinguish against the normal new constructor
        - Add a ua option to allow injection of LWP::UserAgent for testing

        [MOO REWRITE]

lib/Selenium/Firefox.pm
package Selenium::Firefox;
$Selenium::Firefox::VERSION = '1.00';
# ABSTRACT: Use FirefoxDriver without a Selenium server
use Moo;
use Selenium::Firefox::Binary qw/firefox_path/;
use Selenium::CanStartBinary::FindBinary qw/coerce_simple_binary coerce_firefox_binary/;
extends 'Selenium::Remote::Driver';


has '+browser_name' => (
    is => 'ro',
    default => sub { 'firefox' }
);


has 'binary' => (
    is => 'lazy',
    coerce => \&coerce_simple_binary,
    default => sub { 'geckodriver' },
    predicate => 1
);


has 'binary_port' => (
    is => 'lazy',
    default => sub { 9090 }
);


has '_binary_args' => (
    is => 'lazy',
    builder => sub {
        my ($self) = @_;

        if ( $self->marionette_enabled ) {
            my $args = ' --port ' . $self->port
              . ' --marionette-port ' . $self->marionette_binary_port
              . ' --binary "' . $self->firefox_binary . '"';

            return $args;
        }
        else {
            return ' -no-remote';
        }
    }
);

has '+wd_context_prefix' => (
    is => 'ro',
    default => sub {
        my ($self) = @_;

        if ($self->marionette_enabled) {
            return '';
        }
        else {
            return '/hub';
        }

    }
);


has 'marionette_binary_port' => (
    is => 'lazy',
    default => sub { 2828 }
);


has 'marionette_enabled' => (
    is => 'lazy',
    default => 1
);


has 'firefox_binary' => (
    is => 'lazy',
    coerce => \&coerce_firefox_binary,
    predicate => 1,
    builder => 'firefox_path'
);


with 'Selenium::CanStartBinary';


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Selenium::Firefox - Use FirefoxDriver without a Selenium server

=head1 VERSION

version 1.00

=head1 SYNOPSIS

    # These two are the same, and will only work with Firefox 48+
    my $driver = Selenium::Firefox->new;
    my $driver = Selenium::Firefox->new( marionette_enabled => 1 );
    ...
    $driver->shutdown_binary;

    # For Firefox 47 and older, disable marionette:
    my $driver = Selenium::Firefox->new( marionette_enabled => 0 );
    $driver->shutdown_binary;

=head1 DESCRIPTION

B<Breaking Changes:> There are breaking changes in v1.0+ of this
module if you're using it to start FF47; please see L</"Breaking
Changes">. You can ignore this if you're using v1.0+ of this module to
start FF48.

This class allows you to use the FirefoxDriver without needing the JRE
or a selenium server running. Unlike starting up an instance of
S::R::D, do not pass the C<remote_server_addr> and C<port> arguments,
and we will search for the Firefox executable in your $PATH. We'll try
to start the binary, connect to it, and shut it down at the end of the
test.

If the Firefox application is not found in the expected places, we'll
fall back to the default L<Selenium::Remote::Driver> behavior of
assuming defaults of 127.0.0.1:4444 after waiting a few seconds.

If you specify a remote server address, or a port, our assumption is
that you are doing standard S::R::D behavior and we will not attempt
any binary startup.

If you're curious whether your Selenium::Firefox instance is using a
separate Firefox binary, or through the selenium server, you can check
the value of the C<binary_mode> attr after instantiation.

=head1 ATTRIBUTES

=head2 binary

Optional: specify the path to the C<geckodriver> binary - this is NOT
the path to the Firefox browser. To specify the path to your Firefox
browser binary, see the L</firefox_binary> attr.

For Firefox 48 and greater, this is the path to your C<geckodriver>
executable. If you don't specify anything, we'll search for
C<geckodriver> in your $PATH.

For Firefox 47 and older, this attribute does not apply, because the
older FF browsers do not use the separate driver binary startup.

=head2 binary_port

Optional: specify the port that we should bind to. If you don't
specify anything, we'll default to the driver's default port. Since
there's no a priori guarantee that this will be an open port, this is
_not_ necessarily the port that we end up using - if the port here is
already bound, we'll search above it until we find an open one.

See L<Selenium::CanStartBinary/port> for more details, and
L<Selenium::Remote::Driver/port> after instantiation to see what the
actual port turned out to be.

=head2 firefox_profile

Optional: Pass in an instance of L<Selenium::Firefox::Profile>
pre-configured as you please. The preferences you specify will be
merged with the ones necessary for setting up webdriver, and as a
result some options may be overwritten or ignored.

    my $profile = Selenium::Firefox::Profile->new;
    my $firefox = Selenium::Firefox->new(
        firefox_profile => $profile
    );

=head2 marionette_binary_port

Optional: specify the port that we should bind marionette to. If you don't
specify anything, we'll default to the marionette's default port. Since
there's no a priori guarantee that this will be an open port, this is
_not_ necessarily the port that we end up using - if the port here is
already bound, we'll search above it until we find an open one.

    Selenium::Firefox->new(
        marionette_enabled     => 1,
        marionette_binary_port => 12345,
    );

Attempting to specify a C<marionette_binary_port> in conjunction with
setting C<marionette_enabled> does not make sense and will most likely
not do anything useful.

=head2 marionette_enabled

Optional: specify whether
L<marionette|https://developer.mozilla.org/en-US/docs/Mozilla/QA/Marionette>
should be enabled or not. By default, marionette is enabled, which
assumes you are running with Firefox 48 or newer. To use this module
to start Firefox 47 or older, you must pass C<marionette_enabled =>
0>.

    my $ff48 = Selenium::Firefox->new( marionette_enabled => 1 ); # defaults to 1
    my $ff47 = Selenium::Firefox->new( marionette_enabled => 0 );

=head2 firefox_binary

Optional: specify the path to the Firefox browser executable. Although
we will attempt to locate this in your $PATH, you may specify it
explicitly here. Note that path here must point to a file that exists
and is executable, or we will croak.

For Firefox 48 and newer, this will be passed to C<geckodriver> such
that it will attempt to start up the Firefox at the specified path. If
you do not specify anything, we will look for the Firefox browser on
our own in the normal places, but if the browser cannot be found,
we'll probably C<die> during instantiation.

For Firefox 47 and older, this browser path should be the file that we
directly start up.

=head2 custom_args

Optional: specify any additional command line arguments you'd like
invoked during the binary startup. See
L<Selenium::CanStartBinary/custom_args> for more information.

For Firefox 48 and newer, these arguments will be passed to
geckodriver during start up.

For Firefox 47 and older, these arguments will be passed to the
Firefox browser during start up.

=head2 startup_timeout

Optional: specify how long to wait for the binary to start itself and
listen on its port. The default duration is arbitrarily 10 seconds. It
accepts an integer number of seconds to wait: the following will wait
up to 20 seconds:

    Selenium::Firefox->new( startup_timeout => 20 );

See L<Selenium::CanStartBinary/startup_timeout> for more information.

=head1 METHODS

=head2 shutdown_binary

Call this method instead of L<Selenium::Remote::Driver/quit> to ensure
that the binary executable is also closed, instead of simply closing
the browser itself. If the browser is still around, it will call
C<quit> for you. After that, it will try to shutdown the browser
binary by making a GET to /shutdown and on Windows, it will attempt to
do a C<taskkill> on the binary CMD window.

    $self->shutdown_binary;

It doesn't take any arguments, and it doesn't return anything.

We do our best to call this when the C<$driver> option goes out of
scope, but if that happens during global destruction, there's nothing
we can do.

=head1 BREAKING CHANGES

In version v1.0+ and newer, the default behavior is to enable
marionette & geckodriver mode. This means that an existing script that
works with v0.2701 and Firefox v47 will require modification if you
upgrade Selenium::Firefox to v1.0+. That is,

    # v0.2701 of Selenium::Firefox works with FF47 like such; this will not
    # work for FF47 after upgrade:
    my $fx47_old = Selenium::Firefox->new;
    ...
    $fx47_old->shutdown_binary;

    # v1.0 of Selenium::Firefox works with FF47 like this
    my $fx47_new = Selenium::Firefox->new( marionette_enabled => 0);
    ...
    $fx47_new->shutdown_binary;

We default to assuming FF48 and geckodriver mode because all
forthcoming versions of the Firefox browser will be using the
geckodriver architecture, and also because that's consistent with the
rest of the driver setups, which all have separate driver binaries
apart from the browser itself. This means that:

    # v0.2701 of Selenium::Firefox cannot start up FF48 at all

    # v1.0+ of Selenium::Firefox works with FF48+ like this:
    my $fx48 = Selenium::Firefox->new;

As with the other drivers, Selenium::Firefox in marionette/geckodriver
mode requires a `geckodriver` executable in the path or provided
during startup, and it will also attempt to find the path to your
Firefox browser. During testing, we found that it was necessary for us
to pass the Firefox browser file path to the `geckodriver` executable
during start up, or else `geckodriver` would have trouble finding
Firefox.

=head1 SEE ALSO

Please see those modules/websites for more information related to this module.

=over 4

=item *

L<Selenium::Remote::Driver|Selenium::Remote::Driver>

=back

=head1 BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/gempesaw/Selenium-Remote-Driver/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

=head1 AUTHORS

Current Maintainers:

=over 4

=item *

Daniel Gempesaw <gempesaw@gmail.com>

=item *

Emmanuel Peroumalnaïk <peroumalnaik.emmanuel@gmail.com>

=back

Previous maintainers:

=over 4

=item *

Luke Closs <cpan@5thplane.com>

=item *

Mark Stosberg <mark@stosberg.com>

=back

Original authors:

=over 4

=item *

Aditya Ivaturi <ivaturi@gmail.com>

=back

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2010-2011 Aditya Ivaturi, Gordon Child

Copyright (c) 2014-2015 Daniel Gempesaw

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

=cut
